lib/maths/CBoostedTree.cc

```diff
@@ -706,9 +706,23 @@ class CBoostedTree::CImpl final {
         // TODO
     }
 
+    //! Get the number of columns training the model will add to the data frame.
+    std::size_t numberExtraColumnsForTrain() const { return 3; }
+
     //! Get the feature sample probabilities.
     TDoubleVec featureWeights() const { return m_FeatureSampleProbabilities; }
 
+    std::size_t estimateMemoryUsage(std::size_t numberRows, std::size_t numberColumns) const {
+        std::size_t maximumNumberNodes{this->maximumTreeSize(numberRows)};
+        std::size_t forestMemoryUsage{this->m_MaximumNumberTrees * maximumNumberNodes *
+                                      (sizeof(CNode) + sizeof(CLeafNodeStatistics))};
```

Comment:
sizeof(CLeafStatistics) is leaving most of the memory unaccounted. I'd add a static `estimateMemoryUsage` method to `CLeafNodeStatistics` to account for its memory. Note we only maintain statistics for *one* tree, so this shouldn't be multiplied by `m_MaximumNumberTrees`. As well as sizeof(*this), you need to account for:
```
TSizeVec m_FeatureBag;
core::CPackedBitVector m_RowMask;
TDoubleVecVec m_Gradients;
TDoubleVecVec m_Curvatures;
TDoubleVec m_MissingGradients;
TDoubleVec m_MissingCurvatures;
```
`m_FeatureBag` this is size ceil(m_FeatureBagFraction * (numberColumns - 1)). `m_RowMask` the size of this proportional to number of rows, I'd assume close to best case memory usage on average since you're over counting the nodes slightly, so this would be `numberRows / 32`. For `m_Gradients` and `m_Curvatures` we have one vector for each feature, i.e. `numberColumns - 1`, and the vectors are sized based on the number of splits we compute per feature, which is given by `m_NumberSplitsPerFeature`. We have one `m_MissingGradients` and `m_MissingCurvatures` for each feature, so `numberColumns - 1`.

Reply:
Thank you for the details! With all this information it was easy to implement the change.